agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +92,174 @@ class PythonRefAgentConnection : AgentConnection {
-            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
-            if (result != null)
-                receivedMessages.remove(result)
-
-            return result
+    private fun sendJson(obj: Any) = webSocket.sendJson(obj)
+
+    /**
+     * Retrieves a new invite from the agent
+     */
+    override fun generateInvite(): Single<String> {
+        return Single.create { observer ->
+            try {
+                /**
+                 * to generate the invite, send "generate_invite" message to the agent, and wait for "invite_generated"
+                 * which must contain an invite
+                 */
+                webSocket.sendJson(SendMessage(type = MESSAGE_TYPES.GENERATE_INVITE))
+                webSocket.receiveMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).subscribe({ msg ->
+                    observer.onSuccess(msg.invite)
+                }, { e: Throwable -> throw(e) })
+            } catch (e: Throwable) {
+                observer.onError(e)
+            }
         }
     }
 
-    fun waitForMessageOfType(type: String): String {
-        var messageOfType: String? = null
-        while (messageOfType == null) {
-            sleep(500)
-            messageOfType = popMessageOfType(type)
+    /**
+     * Waits for incoming connection from whatever party which possesses the invite
+     */
+    override fun waitForInvitedParty(invite: String): Single<IndyPartyConnection> {
+        return Single.create { observer ->
+            try {
+                if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
+                    /**
+                     * On an incoming connection request, the agent must send the "request_received"
+                     */
+                    webSocket.receiveMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).subscribe({
+                        /**
+                         * on the "request_received" message, reply with "send_response" with remote DID to any incoming connection request.
+                         * at this stage it doesn't matter which party sent the connection request
+                         */
+                        sendJson(RequestSendResponseMessage(it.did))
+                        /**
+                         * Check if the STATE already has the public key ("connection_key") corresponding to the invite
+                         */
+                        sendJson(StateRequest())
+                        webSocket.receiveMessageOfType<ObjectNode>(MESSAGE_TYPES.STATE_RESPONSE).subscribe { stateResponse ->
+                            val pairwise = stateResponse["content"]["pairwise_connections"].find { node ->
+                                node["metadata"]["connection_key"].asText() == getPubkeyFromInvite(invite)
+                            }
+                            if (pairwise != null) {
```

Comment:
What will happen if not my invite is received? 
E.g we have multiple invites in progress at the same time and I received response in StateRequest() from another one.

Reply:
This is exactly what's done in the test. All parties that received invites send "request" message to their agents. Their agents communicate the requests to the endpoint encoded in the invite (the inviting party), the inviting party's agent posts a "request_received" for each "request" to the inviting party. The inviting party replies with a "send_response" for each incoming request, which includes a DID of the requesting party. Until this point the order doesn't matter. The inviting party's agent takes care of pairwise connections and stores all connections' properties in the wallet. When later the inviting party requests the state, it must already have all v_keys and session DIDs set up. Each inviting party's waitForInvitedPary(invite) waits for a proper pairwise connection (matching the public key from the invite) appears in the state. From that point on, the inviting party assumes the connection with a party consumed the given invite is finally set up.